### PR TITLE
Continue PR #341 (Allow `_` in `match` to appear earlier on)

### PR DESCRIPTION
**Pull request by fpoli**
This is the same as https://github.com/lalrpop/lalrpop/pull/341 by @k0pernicus, with an additional commit to test overlapping regular expressions as asked in the review.

(I happened to need exactly this feature :smile:)

### CODE REVIEW

File: lalrpop/src/normalize/token_check/test.rs
Comment (Pat-Lafon) [inline lalrpop/src/normalize/token_check/test.rs:192]:
```suggestion
    assert!(validate_grammar(grammar).is_ok());
```

**Review by yannham — APPROVED** [review]
Many thanks for reviving that old work.

I'm not very knowledgable on this part of the code, but reading the previous issues, PRs, and @nikomatsakis messages, this looks good to me, as far as I can tell.

I'll let other maintainers take a look if they have the time to do so (pinging @youknowone as well), but otherwise, if no one complains, I'll go ahead and merge it at some point.

**Comment by yannham** [comment]
Ah, there is a test failing though. Sounds like the grammar isn't accepted when all default features are disabled. Maybe something related to this unicode issue (e.g. #810)?

File: lalrpop/src/normalize/token_check/test.rs
Comment (fpoli) [inline lalrpop/src/normalize/token_check/test.rs:176]:
```suggestion
// This test requires regex's unicode case support
#[cfg_attr(not(feature = "unicode"), ignore)]
fn match_catch_all_in_first_arm() {
```

**Comment by fpoli** [comment]
> Ah, there is a test failing though. Sounds like the grammar isn't accepted when all default features are disabled. Maybe something related to this unicode issue (e.g. #810)?

Thanks, based on that issue I added `#[cfg_attr(not(feature = "unicode"), ignore)]` and it now works (although I'm not sure why that feature is needed).

**Comment by yannham** [comment]
> (although I'm not sure why that feature is needed)

I believe this is the point of the issue: we shouldn't require unicode by default in the lexer, but unfortunately, we currently do, which is an error. We might probably get rid of the attribute once #810 is closed.

**Comment by youknowone** [comment]
@yannham If this patch itself doesn't have problem, can we merge it first and fix the unicode feature stuff later? Or will it affect to users with/without unicode feature?

**Comment by yannham** [comment]
I think we can just remove the `#[cfg(not...]` now that the unicode part has been fixed, and it should be good to go?

**Comment by youknowone** [comment]
I tried it but didnt work only with it. I didnt look in details.

**Comment by dburgener** [comment]
I believe the issue here is the `\w` character matches unicode "word" characters, not just ASCII ones.

See the regex docs: https://docs.rs/regex/latest/regex/#perl-character-classes-unicode-friendly and the lalrpop docs: https://lalrpop.github.io/lalrpop/lexer_tutorial/001_lexer_gen.html#unicode-compatibility

Replacing "\w" with "[[:word:]]" fixes the issue (I believe "[0-9a-zA-Z_]" would work as well).  Demonstrated here: https://github.com/dburgener/lalrpop/pull/4

I'm happy to either push those two commits onto this branch, or let the original contributor change it.

As a side note, I wonder if we should do something like silently rewrite the unicode character classes to ASCII only ones if the unicode feature isn't enabled.  The fact that using "\s", "\w" "\d" etc implicitly requires unicode support feels pretty counter intuitive to me.

**Comment by dburgener** [comment]
Since I haven't heard anything from the submitter, I went ahead and pushed my fix on this remote.  All the checks are passing now, without the unicode ignore.

**Review by yannham — APPROVED** [review]

**Pull request merged**
